### DISCUSSION TITLE
The expression prohibits newlines

### DISCUSSION THREAD

**Discussion by oovm** [tree-sitter/tree-sitter#1421] [Q&A] [answered]
How to indicate that an expression prohibits `NEWLINES`?

Because most situations do not need to be considered, `WHITESPACE` and `NEWLINE` has already added `extras`.

Then line breaks cannot appear in certain expressions, such as curly braces in `toml` cannot span lines.

And the `return` of javascript which followed by curly braces, whether break the line introduce two different expression trees.

How to deal with these limits?

**Comment by ahlinc** [comment, marked as answer]
For this there are two options:
1. In the `grammar.js` you can try to use `token.immediate()` to imitate such behaviour:
   ```js
   rule1: $ => seq("return", token.immediate(/ +/), token.immediate("{"))
   ```
2. Use an external scanner to check a part of input.

Overall it a known limitation, see #931.
